internal/cmd/io_binary.go

```diff
@@ -0,0 +1,269 @@
+		"CONTAINER_ID=" + id,
+		"CONTAINER_NAMESPACE=" + ns,
+		"CONTAINER_STDOUT=" + stdoutPipe,
+		"CONTAINER_STDERR=" + stderrPipe,
+		"CONTAINER_WAIT=" + waitPipePath,
+	}
+	cmd, err := newBinaryCmd(ctx, uri, envs)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	errCh := make(chan error, 1)
+	// Wait for logging driver to signal to the wait pipe that it's ready to consume IO
+	go func() {
+		b := make([]byte, 1)
+		if _, err := waitPipe.Read(b); err != nil && err != io.EOF {
+			errCh <- err
+			return
+		}
+		errCh <- nil
+	}()
+
+	select {
+	case err = <-errCh:
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to start binary logger")
+		}
+	case <-time.After(binaryCmdStartTimeout):
+		return nil, errors.New("failed to start binary logger: timeout")
+	}
+
+	return &binaryIO{
+		cmd:    cmd,
+		stdout: stdoutPipe,
+		sout:   sout,
+		stderr: stderrPipe,
+		serr:   serr,
+	}, nil
+}
+
+func newBinaryCmd(ctx context.Context, uri *url.URL, envs []string) (*exec.Cmd, error) {
+	if uri.Host == "" && uri.Path == "" {
+		return nil, errors.New("no logging driver path provided")
+	}
+
+	var args []string
+	for k, vs := range uri.Query() {
+		args = append(args, k)
+		if len(vs) > 0 && vs[0] != "" {
+			args = append(args, vs[0])
+		}
+	}
+
+	execPath := uri.Path
+
+	// Absolute path is required, treat "binary://path/to/binary" and "binary:///path/to/binary"
+	// as the same.
+	if uri.Host != "" {
+		execPath = "/" + uri.Host + uri.Path
+	}
+
+	cmd := exec.CommandContext(ctx, execPath, args...)
+	cmd.Env = append(cmd.Env, envs...)
+
+	return cmd, nil
+}
+
+var _ UpstreamIO = &binaryIO{}
+
+// Implements UpstreamIO interface to enable shim pluggable logging
+type binaryIO struct {
+	cmd *exec.Cmd
+
+	binaryCloser sync.Once
+
+	stdin, stdout, stderr string
+
+	sout, serr io.ReadWriteCloser
+	soutCloser sync.Once
+}
+
+// Close named pipes for container stdout and stderr and wait for the binary process to finish.
+func (b *binaryIO) Close(ctx context.Context) {
+	b.soutCloser.Do(func() {
+		if b.sout != nil {
+			err := b.sout.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("error while closing stdout npipe")
+			}
+		}
+		if b.serr != nil {
+			err := b.serr.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("error while closing stderr npipe")
+			}
+		}
+	})
+	b.binaryCloser.Do(func() {
+		done := make(chan error)
+		go func() {
+			done <- b.cmd.Wait()
+		}()
+
+		select {
+		case err := <-done:
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("error while waiting for binary cmd to finish")
+			}
+		case <-time.After(binaryCmdWaitTimeout):
+			log.G(ctx).Errorf("timeout while waiting for binaryIO process to finish. Killing")
+			err := b.cmd.Process.Kill()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("error while killing binaryIO process")
+			}
+		}
+	})
+}
+
+func (b *binaryIO) CloseStdin(ctx context.Context) {
+
+}
```

Comment:
Nit: maybe we shouldn't have space in the function body since this is a noop? Maybe we could do like `ClostStdin(...) {}` or just add a blank `return` inside the body?

Reply:
yeah, sure.